### PR TITLE
[SPARK-9143] [SQL] Add planner rule for automatically inserting Unsafe <-> Safe row format converters

### PR DESCRIPTION
**Pull request by JoshRosen**
Now that we have two different internal row formats, UnsafeRow and the old Java-object-based row format, we end up having to perform conversions between these two formats. These conversions should not be performed by the operators themselves; instead, the planner should be responsible for inserting appropriate format conversions when they are needed.

This patch makes the following changes:
- Add two new physical operators for performing row format conversions, `ConvertToUnsafe` and `ConvertFromUnsafe`.
- Add new methods to `SparkPlan` to allow operators to express whether they output UnsafeRows and whether they can handle safe or unsafe rows as inputs.
- Implement an `EnsureRowFormats` rule to automatically insert converter operators where necessary.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Ah, I guess I should add a test to show that the `Limit` case that I described will work properly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 1]
```diff
@@ -306,6 +306,8 @@ case class UnsafeExternalSort(
   override def output: Seq[Attribute] = child.output
 
   override def outputOrdering: Seq[SortOrder] = sortOrder
+
+  override def outputsUnsafeRows: Boolean = true
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: What about filter?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Yep, meant to change this.  Thanks for reminding me.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Should we also add an assertion to our set operations that the inputs are the same type of row?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: or actually, maybe we can add a general assertion to the `execute` method of `SparkPlan`?

`assert(children.map(_.outputsUnsafeRows).distinct <= 1)`

**Comment by marmbrus** [comment]
LGTM pending filter fix and assertion addition!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 1 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Yeah, let's add it to `execute` I think.  I'll do this shortly.

**Comment by JoshRosen** [comment]
Alright, I've updated this to address your review feedback and also added a lot more assertions and test cases.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/rowFormatConverters.scala [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+          }
+        }
+      } else {
+        operator
+      }
+    case operator: SparkPlan if handlesBothSafeAndUnsafeRows(operator) =>
+      if (operator.children.map(_.outputsUnsafeRows).toSet.size != 1) {
+        // If this operator's children produce both unsafe and safe rows, then convert everything
+        // to safe rows
+        operator.withNewChildren {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/rowFormatConverters.scala]: wouldn't it make more sense to convert to unsafe instead?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/rowFormatConverters.scala]: Yeah, I think so.  I think that choosing to resolve this type of conflict in favor of UnsafeRow should be fine: if unsafe operators are disabled via a feature flag, then the plan shouldn't contain any operators which claim to output unsafe rows so this branch will never be triggered.

I'll update this patch to change this logic.

**Comment by rxin** [comment]
Thanks - merging this in.

**Pull request closed without merging**